### PR TITLE
Check if a locator already exist by name before add it to database [12711]

### PR DESCRIPTION
**Pull request by jparisu**
Monitor fails with assert: `StatisticsParticipantListener.cpp:146 : Assertion locator_ids.empty() || locator_ids.size() == 1' failed.` when it monitors ROS 2 galactic cycloneDDS nodes.

It occurs because the Locatos are created within the Endpoint, and it is possible that they repeat an already existed name in the database.

Signed-off-by: jparisu <javierparis@eprosima.com>

### CODE REVIEW

File: src/cpp/database/database.hpp [hunk 1]
```diff
@@ -797,7 +846,7 @@ class Database
     std::atomic<int64_t> next_id_{0};
 
     //! Read-write synchronization mutex
-    mutable std::shared_timed_mutex mutex_;
+    mutable std::recursive_timed_mutex mutex_;
```
Comment (IkerLuengo) [inline src/cpp/database/database.hpp:849]: why this change? The reason of having a shared mutex is to allow several reading operations on parallel, which the recursive mute does not allow for. Is there any chance that the mutex will be locked several times in the same thread now? I do not see in this pull request any new lock operation that may result in a multiple locking.

**Review by rsanchez15 — APPROVED** [review]

**Review by IkerLuengo — APPROVED** [review]
LGTM
